core/src/main/java/org/apache/calcite/rel/metadata/DelegatingMetadataRel.java

```diff
@@ -0,0 +1,26 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.metadata;
+
+import org.apache.calcite.rel.RelNode;
+
+/**
+ * Interface for {@link RelNode} where the metadata is derived from another node.
+ */
+public interface DelegatingMetadataRel {
```

Comment:
Ideally RelSubset would use this.  However, JaninoRelMetadataProvider explicitly calls out HepRelVertex, so this does not change the exist behavior.  Their is no such exception for RelSubset, so it would result in a change in behavior.  Thus moving RelSubset is worth it own ticker/PR to discuss all the implications of the changes.

Reply:
Can we create a JIRA for it and link it to the original one? Thanks